src/Test/Verilog/Expression.idr

```diff
@@ -0,0 +1,192 @@
+
+public export
+data SVBinaryOperator : (lType : SVType) -> (rType : SVType) -> (retType : SVType) -> Type where
+  SVLogicAnd : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLogicOr : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLogicImply : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLogicalEquiv : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVBiteWiseAnd : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVBiteWiseOr : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVAddBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVAddArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVSubBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVSubArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVMultBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVMultArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVDivBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVDivArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVLessBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLessArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVLessEqBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLessEqArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVGreaterBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVGreaterArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVLogicalEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Logic')
+  SVLogicalEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVLogicalNotEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Logic')
+  SVLogicalNotEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVCaseEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Bit')
+  SVCaseEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Bit')
+  SVCaseNotEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Bit')
+  SVCaseNotEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Bit')
+
+||| Predicate to test if the type is integral as described in the IEEE 1800-2023 standard.
+public export
+data IsIntegralType : SVType -> Type where
+  LogicIsIntegral : IsIntegralType (Var Logic')
+  BitIsIntegral : IsIntegralType (Var Bit')
+  WireIsIntegral : IsIntegralType (Var Wire')
+  UwireIsIntegral : IsIntegralType (Var Uwire')
+  IntIsIntegral : IsIntegralType (Var Int')
+  IntegerIsIntegral : IsIntegralType (Var Integer')
+  UnpackedArrysIsIntegral : IsIntegralType t -> IsIntegralType (Arr (Unpacked t s e))
```

Comment:
I'm not sure that an unpacked array can be an integral type. The IEEE says
```
||| 6.11.1 Integral types
|||
||| The term integral is used throughout this standard to refer to the data types that can represent a single basic
||| integer data type, packed array, packed structure, packed union, or enum.
```

I would use `VarOrPacked` to filter values of packed types